### PR TITLE
fix(multi-tenancy): fix backup/restore after banning namespace

### PR DESCRIPTION
**Pull request by NamanJain8**
This PR fixes bunch of issues in multi-tenancy:

- If the incremental backup was taken after deleting a namespace, the data from the earlier backups were restored. We fix this similar to the way we fix drop operations in backup.
- Mutation on deleted namespace should return, instead of retrying infinitely. This PR fixes it.
- Don't allow deletion of galaxy namespace.
- Extract namespace does not crash the cluster, when JWT is not passed correctly. It returns an error instead.
- Adds a bunch of tests for the same.

EDIT: Removed the fix for drop data as it needs further investigation. 

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7442)

### CODE REVIEW

**Review by ahsanbarkati — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7442)* status: 0 of 33 files reviewed, 12 unresolved discussions (waiting on @manishrjain, @martinmr, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[edgraph/access_ee.go, line 492 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkAv253BtLIDnl_MoI:-MTkAv253BtLIDnl_MoJ:b-aykch8) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/edgraph/access_ee.go#L492)):*
> ```Go
> 	}
> 	ns, err := x.ExtractNamespace(ctx)
> 	x.Check(err)
> ```

Return the error instead of crashing
___
*[edgraph/access_ee.go, line 566 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkB5Cf6shmf1C6CVGP:-MTkB5Cf6shmf1C6CVGQ:b-aykch8) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/edgraph/access_ee.go#L566)):*
> ```Go
> 	}
> 	ns, err := x.ExtractNamespace(ctx)
> 	x.Check(err)
> ```

Return the error instead of crashing
___
*[edgraph/access_ee.go, line 584 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkBVGP1-Gn7Cnp5T0g:-MTkBVGP1-Gn7Cnp5T0h:bpqsnxb) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/edgraph/access_ee.go#L584)):*
> ```Go
> 	aclOp *acl.Operation) (map[string]struct{}, []string) {
> 
> 	// TODO(Naman): Do error handling here.
> ```

We can probably make a struct `AuthPredsResult` containing allowed, blocked preds. And then `authorizePreds()` can return (AuthPredsResult, error)
___
*[systest/backup/multi-tenancy/backup_test.go, line 52 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkDPpU2Ysfb9cVj-eR:-MTkDPpU2Ysfb9cVj-eS:bnn2q34) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/systest/backup/multi-tenancy/backup_test.go#L52)):*
> ```Go
> )
> 
> func sendRestoreRequest(t *testing.T, location string) {
> ```

Can we make these utility function common and reuse them for tests. I see same code lying around in places like `filesystem/backup_test.go`
___
*[systest/backup/multi-tenancy/data/.gitkeep, line 0 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkDn47AkJaVxpD0dXV:-MTkDn480rTUkB1pxgH5:b-t3sofh) ([raw file](https://github.com/dgraph-io/dgraph/blob/832ebb9766f1732186eaecbf3792aff9e3766c6c/systest/backup/multi-tenancy/data/.gitkeep#L0)):*
Probably you want to remove this?
___
*[systest/backup/multi-tenancy/data/backups/.gitkeep, line 0 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkDp375UjW9EoeKazO:-MTkDp375UjW9EoeKazP:b-t3sofh) ([raw file](https://github.com/dgraph-io/dgraph/blob/832ebb9766f1732186eaecbf3792aff9e3766c6c/systest/backup/multi-tenancy/data/backups/.gitkeep#L0)):*
Probably you want to remove this?
___
*[systest/multi-tenancy/basic_test.go, line 41 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkE0WFCsRo8KF_6pzE:-MTkE0WFCsRo8KF_6pzF:bycuxt3) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/systest/multi-tenancy/basic_test.go#L41)):*
> ```Go
> 	galaxyToken := testutil.Login(t,
> 		&testutil.LoginParams{UserID: "groot", Passwd: "password", Namespace: x.GalaxyNamespace})
> 	time.Sleep(1 * time.Second)
> ```

Remove this sleep and retry. Please
___
*[systest/multi-tenancy/basic_test.go, line 116 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkE6Q3Bkt0wjnDFYr4:-MTkE6Q3Bkt0wjnDFYr5:brhumzt) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/systest/multi-tenancy/basic_test.go#L116)):*
> ```Go
> 	galaxyToken := testutil.Login(t,
> 		&testutil.LoginParams{UserID: "groot", Passwd: "password", Namespace: x.GalaxyNamespace})
> 	time.Sleep(1 * time.Second)
> ```

Same here.
___
*[worker/draft.go, line 317 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkEhph2GPgGn_39lvB:-MTkEhph2GPgGn_39lvC:b-vcigh) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/worker/draft.go#L317)):*
> ```Go
> 	if proposal.Mutations.DropOp == pb.Mutations_DATA {
> 		ns, err := strconv.ParseUint(proposal.Mutations.DropValue, 0, 64)
> 		x.Check(err)
> ```

Do we want to crash?
___
*[worker/draft.go, line 784 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkEqSJEYnNeKGYkYyd:-MTkEqSJEYnNeKGYkYye:b-26uw17) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/worker/draft.go#L784)):*
> ```Go
> 		txn.Update()
> 		err := x.RetryUntilSuccess(x.WorkerConfig.MaxRetries, 10*time.Millisecond, func() error {
> 			err := txn.CommitToDisk(writer, commit)
> ```

Can we avoid this nesting?

if err := .. ; err == badger.ErrBannedKey {

}
___
*[worker/restore.go, line 249 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkF2tF-fv6qIsUO_VT:-MTkF2tF-fv6qIsUO_VU:b-vcigh) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/worker/restore.go#L249)):*
> ```Go
> 		case pb.DropOperation_DATA:
> 			ns, err := strconv.ParseUint(operation.DropValue, 0, 64)
> 			x.Check(err)
> ```

Do we want to crash?
___
*[x/x.go, line 265 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkFGTO6kzGeVE_XKw0:-MTkFGTO6kzGeVE_XKw1:b-hl5zyo) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/x/x.go#L265)):*
> ```Go
> // ExtractNamespace parses the namespace value from the incoming gRPC context.
> func ExtractNamespace(ctx context.Context) (uint64, error) {
> 	if !WorkerConfig.AclEnabled {
> ```

Why do we need this? Isn't it that we set `GalaxyNamespace` for all operations if ACL is disabled.

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7442)* status: 0 of 34 files reviewed, 12 unresolved discussions (waiting on @ahsanbarkati, @manishrjain, @martinmr, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[edgraph/access_ee.go, line 492 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkAv253BtLIDnl_MoI:-MTkInJ6C1-kQJ0elYPM:b-n4ruon) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/edgraph/access_ee.go#L492)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Return the error instead of crashing
</blockquote></details>

Done. I made it crash because the descriptions of function says,
```
// upsertGroot must be called after setting the namespace in the context.
```
So I assumed it to be safe to crash.  
___
*[edgraph/access_ee.go, line 566 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkB5Cf6shmf1C6CVGP:-MTkJBhM60ph5HKElkCt:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/edgraph/access_ee.go#L566)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Return the error instead of crashing
</blockquote></details>

Done.
___
*[systest/backup/multi-tenancy/data/.gitkeep, line  at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkDn47AkJaVxpD0dXV:-MTkJwDf95krnN_STuG1:bn9jl3s) ([raw file](https://github.com/dgraph-io/dgraph/blob/832ebb9766f1732186eaecbf3792aff9e3766c6c/systest/backup/multi-tenancy/data/.gitkeep#L)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Probably you want to remove this?
</blockquote></details>

No, they are needed because backup/restore test expect them ot be there. Similar to other backup tests.
___
*[systest/backup/multi-tenancy/data/backups/.gitkeep, line  at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkDp375UjW9EoeKazO:-MTkK5Hy6O5SlsTKKuQe:b2nreh6) ([raw file](https://github.com/dgraph-io/dgraph/blob/832ebb9766f1732186eaecbf3792aff9e3766c6c/systest/backup/multi-tenancy/data/backups/.gitkeep#L)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Probably you want to remove this?
</blockquote></details>

Same as above.
___
*[worker/draft.go, line 317 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkEhph2GPgGn_39lvB:-MTkKblhDfHIeddaMY95:bqcduv8) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/worker/draft.go#L317)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Do we want to crash?
</blockquote></details>

Thanks. We should return an error.
___
*[worker/draft.go, line 784 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkEqSJEYnNeKGYkYyd:-MTkLvKd0W_eWfnZHSht:b-kxijm6) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/worker/draft.go#L784)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Can we avoid this nesting?

if err := .. ; err == badger.ErrBannedKey {

}
</blockquote></details>

I want to scope err outside of if block. So either I declare the variable `var err error` and do what you mentioned. Or the current version.
___
*[worker/restore.go, line 249 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkF2tF-fv6qIsUO_VT:-MTkMMkV1I1QCvcH1ToV:b-t8jn73) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/worker/restore.go#L249)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Do we want to crash?
</blockquote></details>

I think we should crash here. Otherwise, restore will be bring back the data inserted before drop data.
___
*[x/x.go, line 265 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkFGTO6kzGeVE_XKw0:-MTkOiORAjx8we1SO7aF:b-53wp2v) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/x/x.go#L265)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Why do we need this? Isn't it that we set `GalaxyNamespace` for all operations if ACL is disabled.
</blockquote></details>

We aren't attaching the namespace at a couple of places. Like https://github.com/dgraph-io/dgraph/blob/2c7de573090f700841b2d9bda2779c48349fedc5/graphql/admin/schema.go#L85 and https://github.com/dgraph-io/dgraph/blob/2c7de573090f700841b2d9bda2779c48349fedc5/graphql/admin/export.go#L57.

I added this as additional safety. Should we instead attach the namespace at those places instead?
Also, we should mention this behavior in the description of `ExtractNamespace` function then.

**Review by ahsanbarkati — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7442)* status: 0 of 34 files reviewed, 12 unresolved discussions (waiting on @ahsanbarkati, @manishrjain, @martinmr, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[x/x.go, line 265 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkFGTO6kzGeVE_XKw0:-MTkQkAu-dA1ZP4P6K5P:bbp2vlk) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/x/x.go#L265)):*
<details><summary><i>Previously, NamanJain8 (Naman Jain) wrote…</i></summary><blockquote>

We aren't attaching the namespace at a couple of places. Like https://github.com/dgraph-io/dgraph/blob/2c7de573090f700841b2d9bda2779c48349fedc5/graphql/admin/schema.go#L85 and https://github.com/dgraph-io/dgraph/blob/2c7de573090f700841b2d9bda2779c48349fedc5/graphql/admin/export.go#L57.

I added this as additional safety. Should we instead attach the namespace at those places instead?
Also, we should mention this behavior in the description of `ExtractNamespace` function then.
</blockquote></details>

I would want `ExtractNamespace` to just extract otherwise error out. This would help prevent bugs.

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7442)* status: 0 of 36 files reviewed, 12 unresolved discussions (waiting on @ahsanbarkati, @manishrjain, @martinmr, @NamanJain8, @pawanrawal, and @vvbalaji-dgraph)

___
*[systest/backup/multi-tenancy/backup_test.go, line 52 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkDPpU2Ysfb9cVj-eR:-MTl0jTX5xikzX_348Ty:bw9uazo) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/systest/backup/multi-tenancy/backup_test.go#L52)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Can we make these utility function common and reuse them for tests. I see same code lying around in places like `filesystem/backup_test.go`
</blockquote></details>

I have move some of the common part. There is more to refactor. But I think a separate PR can be raised for that.
___
*[x/x.go, line 265 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7442#-MTkFGTO6kzGeVE_XKw0:-MTl4AXu2IffsqDVQllP:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/66f77c77a593d613e7c888735cab1ee6cfd2fa87/x/x.go#L265)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

I would want `ExtractNamespace` to just extract otherwise error out. This would help prevent bugs.
</blockquote></details>

Done.

File: systest/backup/common/utils.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	require.Error(t, result.Err)
+	require.Contains(t, result.Err.Error(), "expected a BackupNum value of 1")
+}
+
+func DirSetup(t *testing.T) {
+	// Clean up data from previous runs.
+	DirCleanup(t)
+
+	for _, dir := range testDirs {
+		if err := os.MkdirAll(dir, os.ModePerm); err != nil {
```
Comment (jarifibrahim) [inline systest/backup/common/utils.go:60]: require.noerror

File: systest/backup/common/utils.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	for _, alpha := range alphaContainers {
+		cmd := []string{"mkdir", "-p", alphaBackupDir}
+		if err := testutil.DockerExec(alpha, cmd...); err != nil {
+			t.Fatalf("Error executing command in docker container: %s", err.Error())
+		}
+	}
+}
+
+func DirCleanup(t *testing.T) {
+	if err := os.RemoveAll(restoreDir); err != nil {
```
Comment (jarifibrahim) [inline systest/backup/common/utils.go:74]: use require.NoError everywhere.

Comment (jarifibrahim) [inline systest/backup/common/utils.go:74]: use require.NoError everywhere.

**Review by jarifibrahim — APPROVED** [review]

File: systest/backup/common/utils.go (hunk 1 above)
Comment (NamanJain8) [reply systest/backup/common/utils.go]: Done all.

**Pull request merged**